feat: transaction caching

wadeking98 opened the pull request:
Added functionality to cache read requests to the ledger.
I've provided added functionality to store the cache in memory or on disk in either a TTL or a LRU style cache. I've left the implementation extendable so users can easily implement storing the cache in a db, redis, whatever.


## Demo: 
The cache will keep values as valid for 24 hours, add the `--use-cache` flag to use the caching functionality. If the cache path isn't supplied then it will use a cache in memory instead of on the file system. If cache size is not specified then it will use a cache with size 1000
`./target/debug/indy-vdr-proxy -g ./genesis.txn --use-cache --cache-size 1024 --cache-path /dev/shm/vdr-cache -p 3000`
### Cache:
![cache](https://github.com/hyperledger/indy-vdr/assets/36937407/7ea9ed27-73bb-4a24-b04c-fe6d5b183ef3)

### No Cache:
![no-cache](https://github.com/hyperledger/indy-vdr/assets/36937407/a748b784-942b-4e31-9ccb-644d8e68c64d)

  
  
With caching enabled on a request to lookup a schema, it takes under 32 milliseconds and is sometimes as fast as 7 milliseconds. Compared with when no cache is enabled, the same request will take over 135 milliseconds to almost a second

andrewwhitehead commented on libindy_vdr/src/pool/runner.rs:
```diff
@@ -50,7 +52,7 @@ impl PoolRunner {
             .unwrap();
             let mut thread = PoolThread::new(pool, receiver);
-            thread.run();
+            thread.run(cache);
```
I think it would make more sense to make cache a property of `PoolThread` and leave `run()` with no arguments?

wadeking98 replied:
Done

andrewwhitehead commented on libindy_vdr/src/pool/runner.rs:
```diff
@@ -217,16 +221,21 @@ impl PoolThread {
+    callback: Callback<RefreshResponse>,
+    cache: Option<Cache<String, (String, RequestResultMeta)>>,
+) {
+    let result = perform_refresh(pool, cache).await;
```
We don't want to use the cache when performing a refresh, it needs to connect to one or more nodes to see if the pool state is up to date.

wadeking98 replied:
Done

andrewwhitehead commented on libindy_vdr/Cargo.toml:
```diff
@@ -60,6 +60,8 @@ thiserror = "1.0"
 time = { version = "=0.3.20", features = ["parsing"] }
 url = "2.2.2"
 zmq = "0.9"
+tokio = { version = "1.35.1", features = ["full"] }
```
async-lock would be a much lighter dependency

wadeking98 replied:
Done

swcurran commented:
For all but RevRegEntries, expiring the cache is not needed since the transactions are immutable. Would it make sense (is it possible?) have a fixed (maximum) sized cache with an LRU expiration policy?

Just as a matter of interest, what is the approach on RevRegEntries for finding cache hits?  It seems unlikely there would be cache hits for RevRegEntry transactions, since the combination of `from`, `to` intervals would have to be the same.

wadeking98 commented:
> For all but RevRegEntries, expiring the cache is not needed since the transactions are immutable. Would it make sense (is it possible?) have a fixed (maximum) sized cache with an LRU expiration policy?
> 
> Just as a matter of interest, what is the approach on RevRegEntries for finding cache hits? It seems unlikely there would be cache hits for RevRegEntry transactions, since the combination of `from`, `to` intervals would have to be the same.

I like the idea of not having a time based expiration, the only thing with the LRU approach is that cache get request will have to mutate the cache as well (to mark an entry as recently used). I don't really have a problem with this but it might look a little weird for end users if the get cache requests mutate things. Still, I think it's a good idea. 

In terms of the revocation registry, I think it's still valid. For example, in the wallet the user may open a proof request multiple times, or select an alternate credential for the proof which would fetch the revocation registry entry. I think the wallet would benefit from having RevRegEntry transactions cached

andrewwhitehead commented:
You could add a TTL and maximum number of items, then prune the cache (removing expired items first) when adding new items.

wadeking98 commented:
I could add support for both and then the end user can decide which one to use, or do you think LRU should be avoided @andrewwhitehead ?

andrewwhitehead commented:
I think the only issue with straight LRU is that commonly used entries might never expire, which means they would never be re-fetched. Also, the get/insert/remove methods of the cache implementation don't necessarily need to take `&mut self`, because locking the Mutex gives you a mutable reference anyway.

swcurran commented:
I think the fetching patterns vary by object type.  Is it worth worrying about that?

Specifically:

- DIDs, ATTRIBs can be “updated” (e.g. a new instance written, and on read the latest is returned) so they do need a “moderate” TTL on them.
  - That said, for DIDs at least, we rarely need them, so reads are likely relatively rare.
- CredDefs and RevRegs don’t ever expire, so can have a very long TTL
- RevRegStatus requests (they aren’t ledger objects) must be cached using their query parameters (from, to) and as such should have a short TTL, as getting them out of the cache fairly quickly is probably a good idea.  Every time a new presentation is generated, the query is likely to be different.

wadeking98 commented:
> I think the only issue with straight LRU is that commonly used entries might never expire, which means they would never be re-fetched. Also, the get/insert/remove methods of the cache implementation don't necessarily need to take `&mut self`, because locking the Mutex gives you a mutable reference anyway.

Got it, I've left the LRU cache there just in case. I've implemented the TTL caching functionality as well. I think that I need to keep insert and remove mutable because I've replaced the mutex on the cache with a RwLock. I only had it as a mutex when the LRU cache was the only implementation we had

wadeking98 commented:
> I think the fetching patterns vary by object type. Is it worth worrying about that?
> 
> Specifically:
> 
>     * DIDs, ATTRIBs can be “updated” (e.g. a new instance written, and on read the latest is returned) so they do need a “moderate” TTL on them.
>       
>       * That said, for DIDs at least, we rarely need them, so reads are likely relatively rare.
> 
>     * CredDefs and RevRegs don’t ever expire, so can have a very long TTL
> 
>     * RevRegStatus requests (they aren’t ledger objects) must be cached using their query parameters (from, to) and as such should have a short TTL, as getting them out of the cache fairly quickly is probably a good idea.  Every time a new presentation is generated, the query is likely to be different.

I think this is a good idea... but I think may be big/complex enough to deserve it's own PR

andrewwhitehead commented:
> I think that I need to keep insert and remove mutable because I've replaced the mutex on the cache with a RwLock

The same is true for RwLock. The contents of an Arc are not generally writable, except in cases like this where a write lock is obtained.

If the TTL varies by transaction type then it might be best for the caller to specify what that is in each request? Although we could also add some defaults.

andrewwhitehead commented:
One important case to test would be a GET_TXN for an undefined (not yet published) sequence number. This can return a response with a 'null' value which we do not want to cache.

wadeking98 commented:
> > I think that I need to keep insert and remove mutable because I've replaced the mutex on the cache with a RwLock
> 
> The same is true for RwLock. The contents of an Arc are not generally writable, except in cases like this where a write lock is obtained.
> 
> If the TTL varies by transaction type then it might be best for the caller to specify what that is in each request? Although we could also add some defaults.

I was thinking the TTL storage type could handle that, maybe in the `new()` function the user passes in a default ttl along with a list of enums that match the request type with an associated type of u64 or something that determines the ttl for that request type

wadeking98 commented:
@swcurran @andrewwhitehead I've added functionality to optionally insert with custom ttl/lru index. The cache itself should be strategy agnostic (ie: doesn't care if lru or ttl is being used), so I had to add each caching strategy to an enum that has their config as an associated type, and then pass in that enum with the ttl/lru value when inserting.  Example here: https://github.com/hyperledger/indy-vdr/blob/8838b0541f49d1d30339341b99b5e14919f899f1/libindy_vdr/src/pool/cache/strategy.rs#L421

wadeking98 commented:
> Cool. I’ll leave it to @andrewwhitehead to review the code.
> 
> What has to be done in AFJ/ACA-Py/Aries VCX to make use of this? Perhaps you could add a ticket to AFJ to outline what is needed, and then I can do the same for ACA-Py?

I will still probably need some help from andrew understanding how to export it to the FFIs and adding it to the wrappers, but after that we'll be ready to use it in afj, aca-py, etc

andrewwhitehead commented on libindy_vdr/src/pool/cache/storage.rs:
```diff
@@ -0,0 +1,219 @@
+    }
+    fn get_key_value(
+        &self,
+        selector: Box<
```
I think this can take an `impl FnOnce` rather than require boxing?

andrewwhitehead commented on indy-vdr-proxy/Cargo.toml:
```diff
@@ -34,6 +34,7 @@ serde_json = "1.0"
 tokio = { version = "1.0", features = ["macros", "rt-multi-thread", "signal"] }
 tokio-rustls = { version = "0.24", optional = true }
 url = "2.2.2"
+sled = "0.34.7"
```
It might be nice to expose a convenience wrapper in the indy_vdr crate so that the proxy doesn't also have to depend directly on sled.

andrewwhitehead commented on libindy_vdr/src/pool/cache/strategy.rs:
```diff
@@ -0,0 +1,442 @@
+        let exp_offset = match config {
+            Some(opt_vec) => opt_vec
+                .iter()
+                .find(|opt| match opt {
```
I think you can use:

```
.find_map(|opt| match opt {
 CacheStrategyConfig::TTL(ttl) => Some(ttl),
 _ => None,
})
```

wadeking98 replied:
I made it accept a list because the config has to be supplied at the cache level which is strategy agnostic. We don't know what cache strategy is being used under the hood and the user may want it to be at the back of the queue if it's using LRU, or they may want it to expire quickly if it's using TTL. In which case they'd supply a config option for both the LRU and TTL caching strategies

andrewwhitehead commented on libindy_vdr/src/pool/helpers.rs:
```diff
@@ -19,12 +20,13 @@ use crate::utils::base58;
 /// Perform a pool ledger status request to see if catchup is required
 pub async fn perform_pool_status_request<T: Pool>(
     pool: &T,
+    cache: Option<Cache<String, (String, RequestResultMeta)>>,
```
No cache needed here

wadeking98 commented:
@andrewwhitehead I've implemented FFI to the best of my knowledge. I saw you implemented a different cache for the pool refresh, I wasn't sure if you wanted the new cache to replace that or not so I kept them separate. I could also use a better name for the FFIs to set the cache right now they're along the lines of `indy_vdr_set_ledger_txn_cache` I wasn't sure if I should change it to something like `indy_vdr_set_runner_cache` or something like that

andrewwhitehead commented on libindy_vdr/src/pool/builder.rs:
```diff
@@ -15,16 +17,22 @@ pub struct PoolBuilder {
+    pub fn new(
+        config: PoolConfig,
+        transactions: PoolTransactions,
+        cache: Option<Cache<String, (String, RequestResultMeta)>>,
```
It's confusing to me how PoolBuilder ignores the cache parameter unless it's used to create a PoolRunner, maybe that should just be a parameter of the into_runner method?

wadeking98 replied:
ya that's a better idea, done

andrewwhitehead commented on libindy_vdr/src/ffi/pool.rs:
```diff
@@ -102,7 +107,7 @@ fn handle_pool_refresh(
         }
         if let Some(new_txns) = new_txns {
-            let runner = PoolBuilder::new(config, new_txns).node_weights(node_weights).refreshed(true).into_runner()?;
+            let runner = PoolBuilder::new(config, new_txns, None).node_weights(node_weights).refreshed(true).into_runner()?;
```
I think this means that after a successful pool refresh (with new pool transactions), the new pool runner will no longer have a cache?

andrewwhitehead commented:
I think we may be better off with one cache strategy rather than TTL and LRU as alternatives. For simplicity, but also because:

1. The cache must always be able to restrict its own size (ie. drop LRU entries) to avoid unlimited growth
2. The cache must support expiry times (TTL) in order to avoid returning long outdated records for mutable transaction types like DIDs

If each LRU cache entry also supports an optional expiry time, then the get() operation can simply return None if the cached entry has expired. We could also do automatic compaction to remove expired entries in one pass.

When fetching a transaction, it should also be possible to pass in a maximum TTL, putting additional bounds on the expiry time of the retrieved record (if any). I don't really see the benefit of overriding the LRU cache position when storing cached records.

wadeking98 commented:
Okay got it, so should I remove the LRU strategy, or should I modify it to support TTL so we'd have a ttl/lru hybrid?

andrewwhitehead commented:
Yes, I think I would modify the LRU one to support automatic expiry (TTL).

wadeking98 commented:
got it, and which strategy should I expose to the FFI? should I expose both?

andrewwhitehead commented:
I don't think we want to support multiple strategies, just the hybrid. @swcurran ?

andrewwhitehead commented on libindy_vdr/src/ffi/pool.rs:
```diff
@@ -40,6 +41,9 @@ pub static POOLS: Lazy<RwLock<BTreeMap<PoolHandle, PoolInstance>>> =
 pub static POOL_CACHE: Lazy<RwLock<Option<Arc<dyn PoolTransactionsCache>>>> =
     Lazy::new(|| RwLock::new(Some(Arc::new(InMemoryCache::new()))));
 
+pub static LEDGER_TXN_CACHE: Lazy<RwLock<Option<Cache<String, (String, RequestResultMeta)>>>> =
```
I think there may be an issue with having one global cache. Clients often connect to multiple ledgers, and the caches need to be distinct, so that a schema ID from one ledger doesn't resolve via the cache when requesting from another.

wadeking98 replied:
the cache key includes the ledger ID:
![image](https://github.com/hyperledger/indy-vdr/assets/36937407/679a740d-4e88-49e3-8777-07f23b9a0116)

wadeking98 replied:
I'm using sha256 to hash the key so we can save space in the cache. We shouldn't have any conflict caching for multiple ledgers in the same cache

andrewwhitehead replied:
I believe that `ledgerId` identifies the plenum 'subledger': 0 is the pool, 1 is the domain ledger for most commonly used transactions, 2 is the config ledger. It is good to include that in the cache key, but it won't differentiate responses from (say) CANdy and Sovrin.

wadeking98 replied:
how did you differentiate between ledgers when you created the pool cache? or did you not need to?

andrewwhitehead replied:
It's not needed there, the set of genesis transactions (the cached values) are enough to differentiate them.

wadeking98 replied:
could we use the pool merkle tree as part of the key? that way it would differentiate between each pool? Having a global cache is nice because then we just have one cache location on the file system, I'd rather not make the user maintain multiple db files for each ledger they use

andrewwhitehead replied:
You could, but then you'd lose the cache when the pool transactions were updated. I think even if you picked like, the first 4 transactions you would risk strange effects when connecting to two ledgers that had forked off of the same base (although that is somewhat theoretical).

wadeking98 replied:
hmm, is there any way to consistently ID a pool across pool refreshes/updates?

wadeking98 replied:
if not I could add a pool id to the type that we could leverage in our caching

wadeking98 commented:
@andrewwhitehead I've merged the TTL and LRU strategies. I kept the old TTL name since it's mostly a TTL cache

andrewwhitehead commented on libindy_vdr/src/pool/cache/strategy.rs:
```diff
@@ -0,0 +1,199 @@
+                cache
+                    .insert("key".to_string(), "value".to_string(), None)
+                    .await;
+                thread::sleep(std::time::Duration::from_millis(1));
```
I'm not sure why these sleeps are necessary, is the cache not updated by the time insert completes?

wadeking98 replied:
They're not really necessary, I just put them there for testing because otherwise all the values in the BtreeMap or sled storage would be keyed under the same millisecond

wadeking98 commented:
@andrewwhitehead I've added functionality to have one cache per ledger to avoid conflicts. If the client connects to two ledgers that have the same initial transactions and uses a filesystem cache then there is still a potential for cache conflicts. But like you said that case may be somewhat theoretical since I'm not sure how many different ledgers are likely to have the same genesis transactions

andrewwhitehead commented on libindy_vdr/src/ffi/mod.rs:
```diff
@@ -72,6 +76,29 @@ pub extern "C" fn indy_vdr_set_cache_directory(path: FfiStr) -> ErrorCode {
+    capacity: usize,
+    expire_offset: u64,
+    path: FfiStr,
+) -> ErrorCode {
```
Could we combine these two methods and just make the path an optional parameter, using the in-memory cache if it's empty or null?

wadeking98 replied:
done

andrewwhitehead commented on libindy_vdr/src/ffi/pool.rs:
```diff
@@ -102,7 +110,9 @@ fn handle_pool_refresh(
         if let Some(new_txns) = new_txns {
-            let runner = PoolBuilder::new(config, new_txns).node_weights(node_weights).refreshed(true).into_runner()?;
+            // set this cache with unique key prefix
+            let txn_cache = read_lock!(LEDGER_CACHE_STRATEGY)?.as_ref().map(|s| Cache::new(s.clone(), init_txns.root_hash_base58().ok()));
```
I believe it should be using the hash of the new transactions here, not the initial ones

wadeking98 replied:
That would change the ID on a pool refresh though wouldn't it? Unless that's what we want. I went with the init_txns assuming those were the genesis transactions and were less likely to change. Additionally the pool cache seems to be using them already as a key so I figured we could use them as a (somewhat) stable key prefix for the txn cache

wadeking98 replied:
I changed it to use the new_txns anyway but I figured I'd ask

andrewwhitehead replied:
ACA-Py does its own caching, so in that case the genesis transactions will change over time anyway. If there are two successful refreshes, then it would also have different 'initial' transactions. I think for consistency and avoiding these edge cases we can just stick with the latest.

andrewwhitehead commented on libindy_vdr/src/pool/cache/mod.rs:
```diff
@@ -0,0 +1,61 @@
+            Some(prefix) => format!("{}{}", prefix, key),
+            None => key.to_string(),
+        }
+    }
```
I think rustfmt needs to be run on this file and the next? It seems odd not having a blank line between methods.

wadeking98 replied:
rust fmt didn't seem to complain, I added the linebreaks anyway though

andrewwhitehead commented on libindy_vdr/src/pool/cache/storage.rs:
```diff
@@ -0,0 +1,227 @@
+/// A hashmap that also maintains a BTreeMap of keys ordered by a given value
+/// This is useful for structures that need fast O(1) lookups, but also need to evict the oldest or least recently used entries
+/// The Ordered Store must contain both the keys and values for persistence
+pub struct OrderedHashMap<K, O, V>(
```
I'm not sure why this is a tuple struct with a 2-tuple as its only member, instead of having two named fields?

andrewwhitehead commented:
Getting pretty close now I think. The solution for handling multiple ledgers should work, it's not the end of the world if the cache gets expired when the pool transactions are changed.

We may want to swap out `sled` later as it seems to be going without maintenance at the moment, but it's fine for now. `fjall` could be a good option? I'd like to find a key-value DB with random element access (or add that feature) so that this could be implemented for the cache eviction: https://danluu.com/2choices-eviction/

wadeking98 commented on wrappers/javascript/indy-vdr-nodejs/src/NodeJSIndyVdr.ts:
```diff
@@ -141,6 +141,11 @@ export class NodeJSIndyVdr implements IndyVdr {
     this.handleError(this.nativeIndyVdr.indy_vdr_set_cache_directory(path))
   }
 
+  public setLedgerTxnCache(options: { capacity: number; expiry_offset_ms: number; path: string }): void {
```
the path argument here could be optional `path?: string` but I'm not sure how well that would cross the Typescript => C => Rust function interfaces. Thought? @andrewwhitehead

berendsliedrecht replied:
undefined will be converted into a nullptr. Should be fine to keep the type at `path?: string`.

andrewwhitehead commented on libindy_vdr/src/ffi/mod.rs:
```diff
@@ -72,6 +77,25 @@ pub extern "C" fn indy_vdr_set_cache_directory(path: FfiStr) -> ErrorCode {
+    catch_err! {
+        debug!("Setting pool ledger transactions cache: capacity={}, expire_offset={}", capacity, expire_offset);
+        let store = match path_opt.as_opt_str() {
+            Some("") => OrderedHashMap::new(new_mem_ordered_store()),
```
```suggestion
            None | Some("") => OrderedHashMap::new(new_mem_ordered_store()),
```
or you could match `path_opt.as_opt_str().unwrap_or_default()`

berendsliedrecht commented on libindy_vdr/include/libindy_vdr.h:
```diff
@@ -481,6 +481,8 @@ ErrorCode indy_vdr_resolve(PoolHandle pool_handle,
 
 ErrorCode indy_vdr_set_cache_directory(FfiStr path);
 
+ErrorCode indy_vdr_set_ledger_txn_cache(size_t capacity, c_ulong expiry_offset_ms, FfiStr path);
```
For all other functions we use `int64_t` or `int32_t`. Could `size_t` (`usize`) and `c_ulong` (`u64`) be switched to `i32` and `i64` respectively?

wadeking98 replied:
we could, but switching to signed integers would cut the maximum capacity and expiry date in half, and it doesn't really make sense to me to support negative capacity or an expiry date that's in the past

andrewwhitehead replied:
Signed numbers are mostly used in the FFI because (supposedly) Java handles those much better than unsigned ones.

wadeking98 replied:
okay got it, I'll change to signed

berendsliedrecht commented on libindy_vdr/src/common/error.rs:
```diff
@@ -130,6 +130,12 @@ impl From<zmq::Error> for VdrError {
     }
 }
 
+impl From<sled::Error> for VdrError {
```
Can leave this as-is for now, but I think we can use `#[from(sled::Error)]` from `thiserror` and it will derive it properly.

berendsliedrecht commented on libindy_vdr/src/pool/cache/mod.rs:
```diff
@@ -0,0 +1,65 @@
+    async fn insert(&mut self, key: K, value: V, custom_exp_offset: Option<u128>) -> Option<V>;
+}
+
+pub struct Cache<K, V> {
```
Would it make sense to require `K` to implement `Display` here as well?

wadeking98 replied:
Sure thing, it's not required by rust, but now that I look it seems like a convention in the codebase

berendsliedrecht commented on libindy_vdr/src/pool/cache/mod.rs:
```diff
@@ -0,0 +1,65 @@
+        self.storage.write().await.remove(&full_key).await
+    }
+
+    pub async fn insert(&self, key: K, value: V, custom_exp_offset: Option<u128>) -> Option<V> {
```
Have never used it myself but do we know the support for `u128`? Is this something that LLVM fixes for us?

wadeking98 replied:
In this thread here: https://stackoverflow.com/questions/57340308/how-does-rusts-128-bit-integer-i128-work-on-a-64-bit-system 
it looks like u128s are compiled down to LLVM and handled by multiple registers

berendsliedrecht commented on libindy_vdr/src/pool/requests/prepared_request.rs:
```diff
@@ -80,6 +81,18 @@ impl PreparedRequest {
+            .as_object_mut()
+            .ok_or_else(|| input_err("Invalid request JSON"))?;
+        req_map.remove("reqId");
+        req_map.remove("signature");
```
This still refers to the `req_json`, right? Otherwise the removes don't do anything as `req_json` is being used after.

wadeking98 replied:
yes req_map is a mutable reference to req_json, but in a slightly different type. `as_object_mut()` returns a mutable reference to req_json in the form of a `&mut Map<String, Value>` object. Insertions and removals made to the map are still reflected in the original object. I think what's happening here is that `as_object_mut()` is returning a reference to the underlying struct that serde_json uses in it's `Value` type

wadeking98 replied:
I've tested this and can confirm that the `reqId` and `signiture` attributes are removed before making the raw key

berendsliedrecht commented on wrappers/javascript/indy-vdr-react-native/cpp/indyVdr.cpp:
```diff
@@ -32,6 +32,16 @@ jsi::Value setCacheDirectory(jsi::Runtime &rt, jsi::Object options) {
+jsi::Value setLedgerTxnCache(jsi::Runtime &rt, jsi::Object options) {
+  auto capacity = jsiToValue<size_t>(rt, options, "capacity");
+  auto expiry_offset_ms = jsiToValue<c_ulong>(rt, options, "expiry_offset_ms");
+  auto path = jsiToValue<std::string>(rt, options, "path");
```
if path is optional the following can be used:

```suggestion
  auto path = jsiToValue<std::string>(rt, options, "path", true);
```

But make sure that before you call `path.c_str()` you first check whether its a nullptr or string. There should be some examples in other funtions to check it from.

berendsliedrecht commented on wrappers/javascript/indy-vdr-react-native/src/NativeBindings.ts:
```diff
@@ -13,6 +13,8 @@ export interface NativeBindings {
 
   setCacheDirectory(options: { path: string }): ReturnObject<never>
 
+  setLedgerTxnCache(options: { capacity: number; expiry_offset_ms: number; path: string }): ReturnObject<never>
```
All the types can be with `path?: string` for the RN wrapper as well.

wadeking98 commented:
I'll need @berendsliedrecht or @andrewwhitehead to merge since I don't have write access to the repo
